README.md
# 桃按钮前端

## 摘要

版本: v0.1.1  
主要工程师: 疯狂的小智  
次要工程师: FinnTenzor  
日期: 2020-06-24  
仓库: [https://gitee.com/echodevelopgroup/echo-button-web-ing](https://gitee.com/echodevelopgroup/echo-button-web-ing)  

## 框架

框架: Vue 2.x  
模板: 官方标准模板  
UI库: element-UI  
其他依赖: mockjs axios  
vue文档: [https://cn.vuejs.org/v2/guide/](https://cn.vuejs.org/v2/guide/)  
vuex文档: [https://vuex.vuejs.org/zh/](https://vuex.vuejs.org/zh/)  
vue-router文档: [https://router.vuejs.org/zh-cn/](https://router.vuejs.org/zh-cn/)  
ElementUI文档: [https://element.eleme.cn/#/zh-CN/component/layout](https://element.eleme.cn/#/zh-CN/component/layout)  
axios文档: [https://www.kancloud.cn/yunye/axios/234845](https://www.kancloud.cn/yunye/axios/234845)  
Mock.js文档: [https://github.com/nuysoft/Mock/wiki](https://github.com/nuysoft/Mock/wiki)  

## 目录结构

1. api 接口相关的函数、拦截器、类型定义等
2. assets LOGO等图片、音频、视频资源 不包括按钮音频
3. components 组件库，组件规范见项目规范
4. layouts 布局库，布局规范见项目规范
5. mock 后端接口模拟，用于后端接口还不能用的时候用来模拟后端返回，写法见官方文档和已写好的部分模拟接口
6. router 前端路由，描述前端所有的页面和定义
7. store 前端全局状态，用于跨页面、跨组件传递状态
8. views 前端各页面
9. 如果某些东西放不到合适的地方，就新建一个文件夹

## 项目规范

__规范是用来提高大家效率的，如果反而降低效率，那么修改调整规范是最好的办法。__

### IDE

使用VSCode或WebStorm或其他

IDE插件：

1. eslint （WS好像自带）
2. editorconfig （WS好像自带）
3. vetur （WS不需要）

### 缩进规范

具体见.editorconfig，主要为2空格缩进和LF换行符

### 代码规范

具体见.eslintrc.js，大部分规范同官方标准，部分特别规则见文件内的设置和注释

### 类型标注

1. 原则上api和util中的所有函数都应该对参数和返回值进行类型标注，类型标注方法见api文件夹内的类型标注。
2. 组件内的函数、方法，若根据组件参数声明可以直接确定，则不必说明类型，否则原则上都应该标注。
3. 上面这个原则，在函数仅被依赖一次，仅被调用一次的时候可以打破。

### 注释

1. 要么不写，要么保证写的注释对其他人有帮助
2. 函数方法上方应该写他的主要功能，`/* */`或`//`都可以
3. 触发性函数、方法 __必须__ 说明触发方法，一般是监听器，触发方法例如`用户按下XX按钮时`
4. 注释引用不要复制，写引用的文件和位置。（防止复制以后，改天别人改了原注释位置导致注释错误，坑别人）
5. 函数体内注释看情况写，简单的不必写，复杂的分块写，不要每行都写。“简单”的标准如下（满足其一）：
    * if/for小于3个
    * 函数功能你能15个字以内说完
    * 变量少于3个
    * 最晚使用的变量距离函数体开头少于10行

### 命名规范

#### html

1. 标签名小写，组件名小写中划线

#### css

1. 类名小写中划线
2. 状态性类名在基础类后加两个下划线，再加状态名，基础类名和状态名都时小写中划线
3. 避免采用复杂的选择器，最好只有类选择器+伪类/伪元素
4. 避免采用容易冲突的命名，最好无scoped全项目类名不冲突

#### js

1. 变量名、函数名、参数名小驼峰
2. 类名大驼峰

#### 文件名

1. vue文件名都采用大驼峰，index除外
2. js文件名都采用小写中划线

### 组件库规范

1. 组件库不嵌套
2. 组件应该是多例的或者未来可能转为多例的
3. 简单组件单文件，复杂组件单个文件夹
4. 如果组件复杂，应该分到其他组件中

### 布局库规范

1. 应该总是仅被router中引用，不要被直接引用

### views

1. 允许嵌套
2. 嵌套结构跟URL的结构走

## 开发和部署

### 开发

```sh
yarn serve
```

开发时反向代理配置在`vue.config.js`中，已经配置好。

### 部署

```sh
yarn build
```

把dist目录复制到网站根目录即可，需要nginx对`/echo-api`和`/echo-static`做反向代理。

## 设计图相关

### 尺寸

1. 设计图PSD水平宽度为4000，页面内部主内容宽度为1110。
2. 对照设计图估计尺寸时，建议在PS里将“图像大小”中的宽度设置为1290，等比缩放，内容部分宽度即为1110。

## 开发历程

1. [x] 播放的页面 FinnTenzor
2. [x] 上传的页面 FinnTenzor
3. [x] 管理页面 FinnTenzor
4. [ ] 序列播放
5. [ ] 自定义播放器
6. [ ] 自定义控制台

## 任务

接任务请把自己的名字写在后面，方便他人确认哪些没人做。

1. [x] 导航栏 FinnTenzor
2. [x] 按钮组 FinnTenzor
3. [x] 上传页面框 FinnTenzor
4. [x] 自定义组件 FinnTenzor
5. [x] 上传审核按钮 FinnTenzor
6. [x] 导航组 FinnTenzor
7. [x] 页面布局 FinnTenzor


CHANGELOG.txt
开始记录时间： 2020-07-13 14:00

2020-07-13 14:00 添加ChangeLog文件
2020-07-13 14:41 编写页面结构部分代码 编写MainLayout.vue等
2020-07-13 14:53 移动阴影代码位置 把之前两个的代码移动到main.css中
2020-07-13 15:39 编写导航栏 除了LOGO可能需要修改 录播站按钮未定外 其他已完成
2020-07-13 16:34 编写导航组 阴影没能实现 其他基本可以了 参数格式估计下个阶段需要修改
2020-07-13 16:48 编写上传和审核按钮
2020-07-13 18:42 编写按钮组外框
2020-07-13 18:44 修复页面中间的阴影
2020-07-13 19:20 编写按钮
2020-07-13 19:22 引入播放控制器 删掉多余的文件
2020-07-14 09:34 编写上传页面
2020-07-14 10:01 修改favicon meta 基路径等
2020-07-14 10:19 补充无JS时的日语提示
2020-07-14 10:24 整理上传审核的按钮的代码 转为一个组件 并添加对新页面打开的支持
2020-07-14 10:31 删除无用代码 添加部分注释
2020-07-14 10:49 修复上传页面上传和提示之间的间距
2020-07-14 11:29 编写页面切换时的一点动效
2020-07-19 11:11 使用vuex控制overview 使用localStorage缓存overview
2020-07-19 11:22 使用import *来解决无法导入类型声明的问题
2020-07-19 11:33 设计前端路由 允许以链接方式和导航互动
2020-07-19 12:43 使用sha1来生成URL 解决upload页面进不去的问题
2020-07-19 12:59 每10个一组显示按钮
2020-07-19 14:05 把切页做成序列 保证切页动作不丢失
2020-07-19 14:15 解决切分类的时候按钮先切换的问题
2020-07-19 14:23 多次切换页面的时候 只保留最后一次切换
2020-07-19 14:36 切换页面时预加载资源
2020-07-19 15:35 允许跳过切页动画 点击导航文本以外的地方也会跳转页面
2020-07-19 15:36 猫耳动画
2020-07-19 16:04 帽子动画
2020-07-19 16:05 Echo叫
2020-07-19 16:46 允许调试时关闭切页动画
2020-07-19 16:56 5%的概率出帽子动画 95%出猫耳动画 导航组排序
2020-07-19 17:47 Alt+C进入配置页面 调整按钮的URL
2020-07-19 17:54 上传页面从分类选取
2020-07-19 19:16 上传功能
2020-07-19 19:17 微调布局 鼠标移动至按钮上显示详情
2020-07-19 20:10 完成审核页面 删除添加注释 调整detail的placeholder
2020-07-19 21:00 修复接口调用错误 审核要求登录
2020-07-19 21:24 修复刷新页面时内容加载错误的问题 播放按钮的时候调用接口
2020-07-19 21:36 把"/"定向到镇站之宝
2020-07-22 01:50 "new"图标的代码 不包括图标
2020-07-29 22:34 完成5级动画并自动调节动画速度
2020-07-30 00:40 修复播放控制的问题
2020-07-30 01:35 关于和友链 右上角按钮敬请期待处理
2020-07-30 15:08 修复ButtonGroup没有margin的问题
2020-07-30 22:34 修复首页白屏问题 改关于 修复导航栏白缝 部署脚本
2020-07-30 22:44 修复部署脚本
2020-08-02 18:19 Safari下禁用预加载 解决按钮垂直间距过近的问题
2020-08-02 18:37 适配1000px~1300px内的设备 移动“关于”代码
2020-08-02 19:05 当前屏幕宽高存入vuex
2020-08-05 00:10 树洞页面(未完成)
2020-08-15 21:11 修复友情链接页面的大量error报错
2020-08-16 11:56 初步完成手机适配
2020-08-16 21:50 添加new按钮
2020-08-16 23:00
1.树洞按钮图标补全 
2.上传页面规则完善
3.顶部导航添加桃子和组的b站空间
2020-08-17 00:13
1.修复部分页面移动适配错误的问题
2.桃树洞页面标题替换
2020-08-17 00:17 修复发送树洞消息没有提示的问题
2020-08-17 19:55 删除前端排序
2020-08-17 20:01
1.手机端允许进入桃树洞
2.防止页脚碰不到页面底端


src/util/sha1.js
/* eslint-disable */
/*
 * A JavaScript implementation of the Secure Hash Algorithm, SHA-1, as defined
 * in FIPS 180-1
 * Version 2.2 Copyright Paul Johnston 2000 - 2009.
 * Other contributors: Greg Holt, Andrew Kepert, Ydnar, Lostinet
 * Distributed under the BSD License
 * See http://pajhome.org.uk/crypt/md5 for details.
 */

/*
 * Configurable variables. You may need to tweak these to be compatible with
 * the server-side, but the defaults work in most cases.
 */
var hexcase = 0;  /* hex output format. 0 - lowercase; 1 - uppercase        */
var b64pad  = ""; /* base-64 pad character. "=" for strict RFC compliance   */

/*
 * These are the functions you'll usually want to call
 * They take string arguments and return either hex or base-64 encoded strings
 */
function hex_sha1(s)    { return rstr2hex(rstr_sha1(str2rstr_utf8(s))); }
function b64_sha1(s)    { return rstr2b64(rstr_sha1(str2rstr_utf8(s))); }
function any_sha1(s, e) { return rstr2any(rstr_sha1(str2rstr_utf8(s)), e); }
function hex_hmac_sha1(k, d)
  { return rstr2hex(rstr_hmac_sha1(str2rstr_utf8(k), str2rstr_utf8(d))); }
function b64_hmac_sha1(k, d)
  { return rstr2b64(rstr_hmac_sha1(str2rstr_utf8(k), str2rstr_utf8(d))); }
function any_hmac_sha1(k, d, e)
  { return rstr2any(rstr_hmac_sha1(str2rstr_utf8(k), str2rstr_utf8(d)), e); }

/*
 * Perform a simple self-test to see if the VM is working
 */
function sha1_vm_test()
{
  return hex_sha1("abc").toLowerCase() == "a9993e364706816aba3e25717850c26c9cd0d89d";
}

/*
 * Calculate the SHA1 of a raw string
 */
function rstr_sha1(s)
{
  return binb2rstr(binb_sha1(rstr2binb(s), s.length * 8));
}

/*
 * Calculate the HMAC-SHA1 of a key and some data (raw strings)
 */
function rstr_hmac_sha1(key, data)
{
  var bkey = rstr2binb(key);
  if(bkey.length > 16) bkey = binb_sha1(bkey, key.length * 8);

  var ipad = Array(16), opad = Array(16);
  for(var i = 0; i < 16; i++)
  {
    ipad[i] = bkey[i] ^ 0x36363636;
    opad[i] = bkey[i] ^ 0x5C5C5C5C;
  }

  var hash = binb_sha1(ipad.concat(rstr2binb(data)), 512 + data.length * 8);
  return binb2rstr(binb_sha1(opad.concat(hash), 512 + 160));
}

/*
 * Convert a raw string to a hex string
 */
function rstr2hex(input)
{
  try { hexcase } catch(e) { hexcase=0; }
  var hex_tab = hexcase ? "0123456789ABCDEF" : "0123456789abcdef";
  var output = "";
  var x;
  for(var i = 0; i < input.length; i++)
  {
    x = input.charCodeAt(i);
    output += hex_tab.charAt((x >>> 4) & 0x0F)
           +  hex_tab.charAt( x        & 0x0F);
  }
  return output;
}

/*
 * Convert a raw string to a base-64 string
 */
function rstr2b64(input)
{
  try { b64pad } catch(e) { b64pad=''; }
  var tab = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
  var output = "";
  var len = input.length;
  for(var i = 0; i < len; i += 3)
  {
    var triplet = (input.charCodeAt(i) << 16)
                | (i + 1 < len ? input.charCodeAt(i+1) << 8 : 0)
                | (i + 2 < len ? input.charCodeAt(i+2)      : 0);
    for(var j = 0; j < 4; j++)
    {
      if(i * 8 + j * 6 > input.length * 8) output += b64pad;
      else output += tab.charAt((triplet >>> 6*(3-j)) & 0x3F);
    }
  }
  return output;
}

/*
 * Convert a raw string to an arbitrary string encoding
 */
function rstr2any(input, encoding)
{
  var divisor = encoding.length;
  var remainders = Array();
  var i, q, x, quotient;

  /* Convert to an array of 16-bit big-endian values, forming the dividend */
  var dividend = Array(Math.ceil(input.length / 2));
  for(i = 0; i < dividend.length; i++)
  {
    dividend[i] = (input.charCodeAt(i * 2) << 8) | input.charCodeAt(i * 2 + 1);
  }

  /*
   * Repeatedly perform a long division. The binary array forms the dividend,
   * the length of the encoding is the divisor. Once computed, the quotient
   * forms the dividend for the next step. We stop when the dividend is zero.
   * All remainders are stored for later use.
   */
  while(dividend.length > 0)
  {
    quotient = Array();
    x = 0;
    for(i = 0; i < dividend.length; i++)
    {
      x = (x << 16) + dividend[i];
      q = Math.floor(x / divisor);
      x -= q * divisor;
      if(quotient.length > 0 || q > 0)
        quotient[quotient.length] = q;
    }
    remainders[remainders.length] = x;
    dividend = quotient;
  }

  /* Convert the remainders to the output string */
  var output = "";
  for(i = remainders.length - 1; i >= 0; i--)
    output += encoding.charAt(remainders[i]);

  /* Append leading zero equivalents */
  var full_length = Math.ceil(input.length * 8 /
                                    (Math.log(encoding.length) / Math.log(2)))
  for(i = output.length; i < full_length; i++)
    output = encoding[0] + output;

  return output;
}

/*
 * Encode a string as utf-8.
 * For efficiency, this assumes the input is valid utf-16.
 */
function str2rstr_utf8(input)
{
  var output = "";
  var i = -1;
  var x, y;

  while(++i < input.length)
  {
    /* Decode utf-16 surrogate pairs */
    x = input.charCodeAt(i);
    y = i + 1 < input.length ? input.charCodeAt(i + 1) : 0;
    if(0xD800 <= x && x <= 0xDBFF && 0xDC00 <= y && y <= 0xDFFF)
    {
      x = 0x10000 + ((x & 0x03FF) << 10) + (y & 0x03FF);
      i++;
    }

    /* Encode output as utf-8 */
    if(x <= 0x7F)
      output += String.fromCharCode(x);
    else if(x <= 0x7FF)
      output += String.fromCharCode(0xC0 | ((x >>> 6 ) & 0x1F),
                                    0x80 | ( x         & 0x3F));
    else if(x <= 0xFFFF)
      output += String.fromCharCode(0xE0 | ((x >>> 12) & 0x0F),
                                    0x80 | ((x >>> 6 ) & 0x3F),
                                    0x80 | ( x         & 0x3F));
    else if(x <= 0x1FFFFF)
      output += String.fromCharCode(0xF0 | ((x >>> 18) & 0x07),
                                    0x80 | ((x >>> 12) & 0x3F),
                                    0x80 | ((x >>> 6 ) & 0x3F),
                                    0x80 | ( x         & 0x3F));
  }
  return output;
}

/*
 * Encode a string as utf-16
 */
function str2rstr_utf16le(input)
{
  var output = "";
  for(var i = 0; i < input.length; i++)
    output += String.fromCharCode( input.charCodeAt(i)        & 0xFF,
                                  (input.charCodeAt(i) >>> 8) & 0xFF);
  return output;
}

function str2rstr_utf16be(input)
{
  var output = "";
  for(var i = 0; i < input.length; i++)
    output += String.fromCharCode((input.charCodeAt(i) >>> 8) & 0xFF,
                                   input.charCodeAt(i)        & 0xFF);
  return output;
}

/*
 * Convert a raw string to an array of big-endian words
 * Characters >255 have their high-byte silently ignored.
 */
function rstr2binb(input)
{
  var output = Array(input.length >> 2);
  for(var i = 0; i < output.length; i++)
    output[i] = 0;
  for(var i = 0; i < input.length * 8; i += 8)
    output[i>>5] |= (input.charCodeAt(i / 8) & 0xFF) << (24 - i % 32);
  return output;
}

/*
 * Convert an array of big-endian words to a string
 */
function binb2rstr(input)
{
  var output = "";
  for(var i = 0; i < input.length * 32; i += 8)
    output += String.fromCharCode((input[i>>5] >>> (24 - i % 32)) & 0xFF);
  return output;
}

/*
 * Calculate the SHA-1 of an array of big-endian words, and a bit length
 */
function binb_sha1(x, len)
{
  /* append padding */
  x[len >> 5] |= 0x80 << (24 - len % 32);
  x[((len + 64 >> 9) << 4) + 15] = len;

  var w = Array(80);
  var a =  1732584193;
  var b = -271733879;
  var c = -1732584194;
  var d =  271733878;
  var e = -1009589776;

  for(var i = 0; i < x.length; i += 16)
  {
    var olda = a;
    var oldb = b;
    var oldc = c;
    var oldd = d;
    var olde = e;

    for(var j = 0; j < 80; j++)
    {
      if(j < 16) w[j] = x[i + j];
      else w[j] = bit_rol(w[j-3] ^ w[j-8] ^ w[j-14] ^ w[j-16], 1);
      var t = safe_add(safe_add(bit_rol(a, 5), sha1_ft(j, b, c, d)),
                       safe_add(safe_add(e, w[j]), sha1_kt(j)));
      e = d;
      d = c;
      c = bit_rol(b, 30);
      b = a;
      a = t;
    }

    a = safe_add(a, olda);
    b = safe_add(b, oldb);
    c = safe_add(c, oldc);
    d = safe_add(d, oldd);
    e = safe_add(e, olde);
  }
  return Array(a, b, c, d, e);

}

/*
 * Perform the appropriate triplet combination function for the current
 * iteration
 */
function sha1_ft(t, b, c, d)
{
  if(t < 20) return (b & c) | ((~b) & d);
  if(t < 40) return b ^ c ^ d;
  if(t < 60) return (b & c) | (b & d) | (c & d);
  return b ^ c ^ d;
}

/*
 * Determine the appropriate additive constant for the current iteration
 */
function sha1_kt(t)
{
  return (t < 20) ?  1518500249 : (t < 40) ?  1859775393 :
         (t < 60) ? -1894007588 : -899497514;
}

/*
 * Add integers, wrapping at 2^32. This uses 16-bit operations internally
 * to work around bugs in some JS interpreters.
 */
function safe_add(x, y)
{
  var lsw = (x & 0xFFFF) + (y & 0xFFFF);
  var msw = (x >> 16) + (y >> 16) + (lsw >> 16);
  return (msw << 16) | (lsw & 0xFFFF);
}

/*
 * Bitwise rotate a 32-bit number to the left.
 */
function bit_rol(num, cnt)
{
  return (num << cnt) | (num >>> (32 - cnt));
}

/**
 * 计算输入字符串的sha1值
 * @param {string} input 输入串
 * @returns {string} hash结果
 */
export function sha1(input) {
  return hex_sha1(input).substr(0, 8)
}
